numba/tests/test_stencils.py

```diff
@@ -365,12 +407,1272 @@ def test_impl(n):
+        a = np.arange(12).reshape(3, 4)
+
+        def kernel(a):
+            return a[0, 0] + a[0, 0] + 1.
+        self.check(kernel, a)
+
+    def test_basic22(self):
+        """rel idx const expr folding, add const"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            return a[1 + 0, 0] + a[0, 0] + 1.
+        self.check(kernel, a)
+
+    def test_basic23(self):
+        """rel idx, work in body"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            x = np.sin(10 + a[2, 1])
+            return a[1 + 0, 0] + a[0, 0] + x
+        self.check(kernel, a)
+
+    def test_basic23a(self):
+        """rel idx, dead code should not impact rel idx"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            x = np.sin(10 + a[2, 1])
+            return a[1 + 0, 0] + a[0, 0]
+        self.check(kernel, a)
+
+    def test_basic24(self):
+        """1d idx on 2d arr"""
+        a = np.arange(12).reshape(3, 4)
+
+        def kernel(a):
+            return a[0] + 1.
+        self.check(kernel, a, expected_exception=[ValueError, TypingError])
+
+    # Should this work at all? does it make sense?
+    # pyStencil raises as there's no rel indices
+    def test_basic25(self):
+        """1d idx on 2d arr"""
+        a = np.arange(12).reshape(3, 4)
+
+        def kernel(a):
+            return 1.
+        self.check(kernel, a, expected_exception=[ValueError])
+
+    def test_basic26(self):
+        """3d arr"""
+        a = np.arange(64).reshape(4, 8, 2)
+
+        def kernel(a):
+            return a[0, 0, 0] - a[0, 1, 0] + 1.
+        self.check(kernel, a)
+
+    def test_basic27(self):
+        """4d arr"""
+        a = np.arange(128).reshape(4, 8, 2, 2)
+
+        def kernel(a):
+            return a[0, 0, 0, 0] - a[0, 1, 0, -1] + 1.
+        self.check(kernel, a)
+
+    def test_basic28(self):
+        """type widen """
+        a = np.arange(12).reshape(3, 4).astype(np.float32)
+
+        def kernel(a):
+            return a[0, 0] + np.float64(10.)
+        self.check(kernel, a)
+
+    def test_basic29(self):
+        """const index from func """
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            return a[0, int(np.cos(0))]
+        self.check(kernel, a, expected_exception=[ValueError, LoweringError])
+
+    def test_basic30(self):
+        """signed zeros"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            return a[-0, -0]
+        self.check(kernel, a)
+
+    # this passes parfors, fails in pyStencil
+    def test_basic31(self):
+        """does a const propagate?"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            t = 1
+            return a[t, 0]
+        self.check(kernel, a)
+        
+    # this fails all, need to decide whether to entirely support const prop + fold
+    # or not.
+    def test_basic31b(self):
+        """does a const propagate?"""
```

Comment:
It seems like one level of constant propagation works (as in `test_basice31`), but that is all, is this intended? Is it relatively quick to support it fully? If not perhaps we document it as such and defer the work to another ticket?

Reply:
I believe that constant propagation happens in the context of generic copy propagation.  What you are describing here is constant folding and that isn't implemented.  I believe this one should be deferred to a separate ticket.

Reply:
Indeed, comment https://github.com/IntelLabs/numba/pull/37/files/3e80e3f112e52e821772b75e3b897edfb9b59f71#diff-9e13229a95b625e1c7496892f1880081R1363 has `prop + fold` but the docstring/other comment doesn't, this is an oversight on my part. Copy propagation is definitely working, I definitely think constant folding should be deferred as it is nice-to-have but not essential for this feature, I'll open a ticket.